doc/guides/modules.rst

```diff
@@ -253,9 +263,43 @@ Where 23 in the example above indicated the pull request number submitted to the
 *my_module* repository. Once the module changes are reviewed and merged, the
 revision needs to be changed to the commit hash from the module repository.
 
-
+Upmerging forks
+===============
+
+When a module is in fact a fork of an existing, external project that is hosted
+on GitHub or elsewhere, it is often necessary to upmerge it: update the copy
+that is hosted under the `zephyrproject-rtos`_ GitHub organization with the
+latest changes from the external upstream repository.
+
+The procedure for upmerging module repositories that are forks of external
+projects is constrained by the following two factors:
+
+* History must **never** be rewritten in any repository that is referenced in
+  the `west.yml`_ manifest file, since that could break older revisions of the
+  repository set as a whole. This therefore excludes rebasing the module
+  repository and keeping the zephyr-specific changes always on top.
+* GitHub does not deal correctly with merge commits contained inside Pull
+  Requests, which prevents us from using regular ``git merge`` operations
+  without having to resort to pushing directly to the repository.
+
+Due to the above, this is the procedure that you must use in order to upmerge a
+module repository with the latest changes from the external upstream
+repository:
+
+#. Make a note of the current upstream revision that the module repository is
+   currently upmerged to.
+#. Take the changes from the current to the latest upstream revision and apply
```

Comment:
This might be controversial, but it's the only way I can think of to avoid the usual upmerge pitfalls.
Perhaps we should add a templated commit message here that says:

```
upmerge: Update from SHA1 to SHA2

Origin: project@SHA2
```

Reply:
The other option is to allow multiple upmerge strategies

Reply:
To say publicly what I've said before in slack/telecons: When I'm working on forks like those Zephyr requires I usually use a workspace with access to every upstream source.  I find it extremely inconvenient to be guessing what upstream commit a particular fork commit is supposed to be equivalent to.  A comment in a commit would at least provide a clue, but allowing real merge commits where you can see the upstream tag in its canonical form associated with the real SHA1 is the only sane way to be doing cross-repository comparisons.

Reply:
I agree that real merge commits (if merging) is the sane way to handle it.

I still think we should consider allowing sporadic rebasing, but that's a different question.